client/hostvolumemanager/host_volumes_test.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package hostvolumemanager
+
+import (
+	"errors"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper/testlog"
+	"github.com/shoenig/test/must"
+)
+
+func TestNewHostVolumeManager_restoreState(t *testing.T) {
+	log := testlog.HCLogger(t)
+	vol := &HostVolumeState{
+		ID: "test-vol-id",
+		CreateReq: &cstructs.ClientHostVolumeCreateRequest{
+			ID:       "test-vol-id",
+			PluginID: "mkdir",
+		},
+	}
+
+	t.Run("happy", func(t *testing.T) {
+		// put our volume in state
+		state := newMockState()
+		must.NoError(t, state.PutDynamicHostVolume(vol))
+
+		// new volume manager should load it from state and run Create,
+		// resulting in a volume directory in this mountDir.
+		mountDir := t.TempDir()
+
+		_, err := NewHostVolumeManager(log, state, time.Second, "/wherever", mountDir)
+		must.NoError(t, err)
+
+		volPath := filepath.Join(mountDir, vol.ID)
+		must.DirExists(t, volPath)
+	})
+
+	t.Run("get error", func(t *testing.T) {
+		state := newMockState()
+		state.getErr = errors.New("get error")
+		_, err := NewHostVolumeManager(log, state, time.Second, "/wherever", "/wherever")
+		// error loading state should break the world
+		must.ErrorIs(t, err, state.getErr)
+	})
+
+	t.Run("put error", func(t *testing.T) {
+		state := newMockState()
+		// add the volume to try to restore
+		must.NoError(t, state.PutDynamicHostVolume(vol))
+
+		// make sure mock state actually errors, to ensure we're ignoring it
+		state.putErr = errors.New("put error")
+		must.Error(t, state.PutDynamicHostVolume(vol))
+
+		// restore process also attempts a state Put during vol Create
+		_, err := NewHostVolumeManager(log, state, time.Second, "/wherever", t.TempDir())
+		// but it should not cause the whole agent to fail to start
+		must.NoError(t, err)
+	})
+}
+
+var _ HostVolumeStateManager = &mockState{}
+
+func newMockState() *mockState {
+	return &mockState{
+		vols: make(map[string]*HostVolumeState),
+	}
+}
```

Comment:
The `client/state/db_mem.go` and `client/state/db_error.go` files are intended to be mock client states you can use for testing these scenarios.

Reply:
I got an import cycle trying to use them, because they import `hostvolumemanager` for its `HostVolumeState`.  would it make sense to move `HostVolumeState` into `client/structs`? I can avoid the import cycle that way.